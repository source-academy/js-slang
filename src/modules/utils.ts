/**
 * For tabs that are export default declarations, we need to remove
 * the `export default` bit from the front before they can be loaded
 * by js-slang
 */
export function evalRawTab(text: string) {
  if (text.startsWith('export default')) {
    text = text.substring(14)
  }

  return eval(text)
}
