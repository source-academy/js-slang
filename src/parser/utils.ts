import {
  Comment,
  ecmaVersion,
  Node,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { Program, SourceLocation } from 'estree'

import { Context } from '..'
import { DEFAULT_ECMA_VERSION } from '../constants'
import { SourceError } from '../types'
import { validateAndAnnotate } from '../validator/validator'
import { MissingSemicolonError, TrailingCommaError } from './errors'
import { AcornOptions, BabelOptions } from './types'

/**
 * Generates options object for acorn parser
 *
 * @param ecmaVersion ECMA version
 * @param errors error container
 * @param throwOnError throw on error if true else push to error container and resume exec
 * @param options partial acorn options
 * @returns
 */
export const createAcornParserOptions = (
  ecmaVersion: ecmaVersion,
  errors?: SourceError[],
  options?: Partial<AcornOptions>,
  throwOnError?: boolean
): AcornOptions => ({
  ecmaVersion,
  sourceType: 'module',
  locations: true,
  onInsertedSemicolon(_tokenEndPos: number, tokenPos: Position) {
    const error = new MissingSemicolonError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  onTrailingComma(_tokenEndPos: number, tokenPos: Position) {
    const error = new TrailingCommaError(positionToSourceLocation(tokenPos, options?.sourceFile))
    if (throwOnError) throw error
    errors?.push(error)
  },
  ...options
})

/**
 * Parses a single expression at a specified offset
 *
 * @param programStr program string
 * @param offset position offset
 * @param ecmaVersion ECMA version
 * @returns acorn AST Node if parse succeeds else null
 */
export function parseAt(
  programStr: string,
  offset: number,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): Node | null {
  try {
    return acornParseAt(programStr, offset, { ecmaVersion })
  } catch (_error) {
    return null
  }
}

/**
 * Parse a program, returning alongside comments found within that program
 *
 * @param programStr program string
 * @param ecmaVersion ECMA version
 * @returns tuple consisting of the parsed program, and a list of comments found within the program string
 */
export function parseWithComments(
  programStr: string,
  ecmaVersion: ecmaVersion = DEFAULT_ECMA_VERSION
): [Program, Comment[]] {
  let comments: acorn.Comment[] = []
  const acornOptions: AcornOptions = createAcornParserOptions(
    ecmaVersion,
    undefined,
    {
      onComment: comments
    },
    undefined
  )

  let ast: Program | undefined
  try {
    ast = acornParse(programStr, acornOptions) as unknown as Program
  } catch {
    comments = []
    ast = acornLooseParse(programStr, acornOptions)
  }

  return [ast, comments]
}

/**
 * Parse program with error-tolerant acorn parser
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function looseParse(programStr: string, context: Context): Program {
  return acornLooseParse(
    programStr,
    createAcornParserOptions(DEFAULT_ECMA_VERSION, context.errors)
  ) as unknown as Program
}

/**
 * TODO
 *
 * @param programStr program string
 * @param context js-slang context
 * @returns ast for program string
 */
export function typedParse(programStr: string, context: Context): Program {
  const ast: Program = looseParse(programStr, context)
  return validateAndAnnotate(ast, context)
}

/**
 * Converts acorn parser Position object to SourceLocation object
 *
 * @param position acorn Position object
 * @returns SourceLocation
 */
export const positionToSourceLocation = (position: Position, source?: string): SourceLocation => ({
  start: { ...position },
  end: { ...position, column: position.column + 1 },
  source
})

export const defaultBabelOptions: BabelOptions = {
  sourceType: 'module',
  plugins: ['typescript', 'estree']
}
