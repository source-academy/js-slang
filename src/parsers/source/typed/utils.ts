import * as es from 'estree'

/**
 * Transforms an AST generated by the Babel parser into one
 * that is compliant with the ESTree types. While the Babel
 * parser-generated AST is mostly compliant with the ESTree
 * specifications due to the use of the 'estree' plugin, not
 * everything is compliant with the ESTree types.
 *
 * @param program The AST to be transformed.
 */
export const transformBabelASTToESTreeCompliantAST = (program: es.Program): void => {
  renameFilenameAttributeToSource(program)
}

/**
 * Renames the 'filename' attribute on AST nodes to 'source'.
 * This is because the Babel parser tags AST nodes with the
 * 'filename' attribute when the name of the file in which
 * the AST was generated from is given. However, by the ESTree
 * types that js-slang uses, this attribute should be named
 * 'source' instead of 'filename'. As a workaround, we
 * recursively walk the AST generated by the Babel parser and
 * replace all instances of the 'filename' attribute with
 * 'source'.
 *
 * Note that we do not use acorn-walk here as a precondition
 * of acorn-walk is that the AST is compliant with the ESTree
 * specifications. Since the reason why we are transforming
 * the AST generated by the Babel parser is because it is not
 * entirely compliant with the ESTree specifications in the
 * first place, we avoid the use of acorn-walk.
 *
 * @param node The node to be transformed.
 */
const renameFilenameAttributeToSource = (node: Record<string, any>): void => {
  // Rename all 'filename' attributes to 'source'.
  if (node.hasOwnProperty('filename')) {
    node.source = node.filename
    delete node.filename
  }
  // Iterate over all property values of the AST node.
  Object.values(node).forEach(value => {
    // If the value is null or undefined, do nothing.
    if (value === null || value === undefined) {
      return
    }
    // If the value is an array, check if any of its elements
    // are objects (i.e., AST nodes) which we should recurse on.
    if (Array.isArray(value)) {
      value.forEach(element => {
        if (element === null || element === undefined) {
          return
        }
        if (typeof element !== 'object') {
          return
        }
        renameFilenameAttributeToSource(element)
      })
    }
    // If the value is a primitive type, do nothing.
    if (typeof value !== 'object') {
      return
    }
    // Otherwise, the value is an object (i.e., AST node) which
    // we recurse on.
    renameFilenameAttributeToSource(value)
  })
}
