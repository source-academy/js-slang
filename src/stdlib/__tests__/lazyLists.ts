import { stripIndent } from '../../utils/formatters'
import { expectParsedError, expectResult } from '../../utils/testing'

test('pair creates pair', () => {
  return expectResult(
    stripIndent`
    is_pair (pair(1, 'a string ""'));
    `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('head works', () => {
  return expectResult(
    stripIndent`
    head(pair(1, 'a string ""'));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('tail works', () => {
  return expectResult(
    stripIndent`
    tail(pair(1, 'a string ""'));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`"a string \\"\\""`)
})

test('tail of a 1 element list is null', () => {
  return expectResult(
    stripIndent`
    tail(list(1));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`null`)
})

test('empty list is null', () => {
  return expectResult(
    stripIndent`
    list();
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot('null')
})

test('for_each', () => {
  return expectResult(
    stripIndent`
    let sum = 0;
    for_each(x => {
      sum = sum + x;
    }, list(1, 2, 3));
    sum;
  `,
    { chapter: 3, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`6`)
})

test('map', () => {
  return expectResult(
    stripIndent`
    equal(map(x => 2 * x, list(12, 11, 3)), list(24, 22, 6));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('filter', () => {
  return expectResult(
    stripIndent`
    equal(filter(x => x <= 4, list(2, 10, 1000, 1, 3, 100, 4, 5, 2, 1000)), list(2, 1, 3, 4, 2));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('build_list', () => {
  return expectResult(
    stripIndent`
    equal(build_list(5, x => x * x), list(0, 1, 4, 9, 16));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('reverse', () => {
  return expectResult(
    stripIndent`
    equal(reverse(list("string", null, undefined, null, 123)), list(123, null, undefined, null, "string"));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('append', () => {
  return expectResult(
    stripIndent`
    equal(append(list("string", 123), list(456, null, undefined)), list("string", 123, 456, null, undefined));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('member', () => {
  return expectResult(
    stripIndent`
    equal(
      member("string", list(1, 2, 3, "string", 123, 456, null, undefined)),
      list("string", 123, 456, null, undefined));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('remove', () => {
  return expectResult(
    stripIndent`
    remove(1, list(1));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`null`)
})

test('remove not found', () => {
  return expectResult(
    stripIndent`
    equal (remove(2, list(1)),list(1));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('remove_all', () => {
  return expectResult(
    stripIndent`
    equal(remove_all(1, list(1, 2, 3, 4, 1, 1, "1", 5, 1, 1, 6)), list(2, 3, 4, "1", 5, 6));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('remove_all not found', () => {
  return expectResult(
    stripIndent`
    equal(remove_all(1, list(2, 3, "1")), list(2, 3, "1"));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('enum_list', () => {
  return expectResult(
    stripIndent`
    equal(enum_list(1, 5), list(1, 2, 3, 4, 5));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('enum_list with floats', () => {
  return expectResult(
    stripIndent`
    equal(enum_list(1.5, 5), list(1.5, 2.5, 3.5, 4.5));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})

test('list_ref', () => {
  return expectResult(
    stripIndent`
    list_ref(list(1, 2, 3, "4", 4), 4);
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`4`)
})

test('accumulate', () => {
  return expectResult(
    stripIndent`
    accumulate((curr, acc) => curr + acc, 0, list(2, 3, 4, 1));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`10`)
})

test('list_to_string', () => {
  return expectResult(
    stripIndent`
    list_to_string(list(1, 2, 3));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`"[1,[2,[3,null]]]"`)
})
test('bad number error build_list', () => {
  return expectParsedError(
    stripIndent`
        build_list('1', x => x);
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`"Line 45: Expected number on left hand side of operation, got string."`)
})

test('bad number error enum_list', () => {
  return expectParsedError(
    stripIndent`
enum_list('1', '5');
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 139: Expected string on right hand side of operation, got number."`
  )
})

test('bad number error enum_list', () => {
  return expectParsedError(
    stripIndent`
enum_list('1', 5);
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 139: Expected string on right hand side of operation, got number."`
  )
})

test('bad number error enum_list', () => {
  return expectParsedError(
    stripIndent`
enum_list(1, '5');
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 140: Expected number on right hand side of operation, got string."`
  )
})

test('bad index error list_ref', () => {
  return expectParsedError(
    stripIndent`
list_ref(list(1, 2, 3), 3);
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 148: Error: tail(xs) expects a pair as argument xs, but encountered null"`
  )
})

test('bad index error list_ref', () => {
  return expectParsedError(
    stripIndent`
list_ref(list(1, 2, 3), -1);
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 148: Error: tail(xs) expects a pair as argument xs, but encountered null"`
  )
})

test('bad index error list_ref', () => {
  return expectParsedError(
    stripIndent`
list_ref(list(1, 2, 3), 1.5);
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(
    `"Line 148: Error: tail(xs) expects a pair as argument xs, but encountered null"`
  )
})

test('bad index error list_ref', () => {
  return expectParsedError(
    stripIndent`
list_ref(list(1, 2, 3), '1');
`,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`"Line 149: Expected number on left hand side of operation, got string."`)
})

test('arguments are not evaluated for pair', () => {
  return expectResult(
    stripIndent`
    head(pair(1,head(null)));
  `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})
test('arguments are not evaluated for list', () => {
  return expectResult(
    stripIndent`
     head(list(1,head(null)));
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('recursive pair definitions are possible (tail)', () => {
  return expectResult(
    stripIndent`
     const a = pair (1,a);
     head(a) + head(tail(a));
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('recursive pair definitions are possible (head)', () => {
  return expectResult(
    stripIndent`
     const a = pair (a,1);
     tail(a) + tail(head(a));
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('recursive list definitions are possible (head)', () => {
  return expectResult(
    stripIndent`
     const a = list (1,a);
     head(a) + head(head(tail(a)));
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('is_list on infinite lists works', () => {
  return expectResult(
    stripIndent`
     const a = list(1,a);
     is_list(a);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`true`)
})
test('list_ref on infinite lists', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     list_ref(a,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('map on infinite lists works', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     const b = map(x => 2 * x, a);
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('map on infinite lists works', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     const b = map(x => 2 * x, a);
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('append left list is infinite', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     const b = append(a, list(3,4));
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('append right list is infinite', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     const b = append(list(3,4),a);
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('remove on infinite list', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,a);
     const b = remove(1,a);
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`1`)
})

test('remove all ones on infinite list of ones and twos', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,pair(2,a));
     const b = remove_all(1,a);
     list_ref(b,200);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})

test('filter on infinite lists', () => {
  return expectResult(
    stripIndent`
     const a = pair(1,pair(2,a));
     const b = filter(x => x % 2 === 0,a);
     list_ref(b,1);
 `,
    { chapter: 2, native: true, variant: 'lazy' }
  ).toMatchInlineSnapshot(`2`)
})
