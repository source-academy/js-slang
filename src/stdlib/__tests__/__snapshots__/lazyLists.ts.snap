// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accumulate: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "accumulate((curr, acc) => curr + acc, 0, list(2, 3, 4, 1));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`append left list is infinite: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
const b = append(a, list(3,4));
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`append right list is infinite: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
const b = append(list(3,4),a);
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`append: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(append(list(123, 123), list(456, 456, 456)), list(123, 123, 456, 456, 456));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`arguments are not evaluated for list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(list(1,head(null)));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`arguments are not evaluated for pair: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(pair(1,head(null)));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`bad index error list_ref: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "list_ref(list(1, 2, 3), 3);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 158: Error: tail(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad index error list_ref: expectParsedError 2`] = `
Object {
  "alertResult": Array [],
  "code": "list_ref(list(1, 2, 3), -1);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 158: Error: tail(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad index error list_ref: expectParsedError 3`] = `
Object {
  "alertResult": Array [],
  "code": "list_ref(list(1, 2, 3), 1.5);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 158: Error: tail(xs) expects a pair as argument xs, but encountered null",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad index error list_ref: expectParsedError 4`] = `
Object {
  "alertResult": Array [],
  "code": "list_ref(list(1, 2, 3), '1');",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 159: Expected string on right hand side of operation, got number.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad number error build_list: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "build_list(x => x, '1');",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 55: Expected number on left hand side of operation, got string.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad number error enum_list: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "enum_list('1', '5');",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 149: Expected string on right hand side of operation, got number.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad number error enum_list: expectParsedError 2`] = `
Object {
  "alertResult": Array [],
  "code": "enum_list('1', 5);",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 149: Expected string on right hand side of operation, got number.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`bad number error enum_list: expectParsedError 3`] = `
Object {
  "alertResult": Array [],
  "code": "enum_list(1, '5');",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 150: Expected number on right hand side of operation, got string.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`build_list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(build_list(x => x * x, 5), list(0, 1, 4, 9, 16));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`empty list is null: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "list();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`enum_list with floats: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(enum_list(1.5, 5), list(1.5, 2.5, 3.5, 4.5));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`enum_list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(enum_list(1, 5), list(1, 2, 3, 4, 5));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`filter on infinite lists: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,pair(2,a));
const b = filter(x => x % 2 === 0,a);
list_ref(b,1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`filter: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(filter(x => x <= 4, list(2, 10, 1000, 1, 3, 100, 4, 5, 2, 1000)), list(2, 1, 3, 4, 2));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`for_each: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "let sum = 0;
for_each(x => {
  sum = sum + x;
}, list(1, 2, 3));
sum;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 6,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`head works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(pair(1, 'a string \\"\\"'));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`is_list on infinite lists works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = list(1,a);
is_list(a);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`list_ref on infinite lists: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
list_ref(a,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`list_ref: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "list_ref(list(1, 2, 3, \\"4\\", 4), 4);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 4,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`list_to_string: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "list_to_string(list(1, 2, 3));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "[1,[2,[3,null]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map on infinite lists works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
const b = map(x => 2 * x, a);
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map on infinite lists works: expectResult 2`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
const b = map(x => 2 * x, a);
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(map(x => 2 * x, list(12, 11, 3)), list(24, 22, 6));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`member: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(
  member(4, list(1, 2, 3, 4, 123, 456, 789)),
  list(4, 123, 456, 789));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`pair creates pair: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "is_pair (pair(1, 'a string \\"\\"'));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`recursive list definitions are possible (head): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = list (1,a);
head(a) + head(head(tail(a)));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`recursive pair definitions are possible (head): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair (a,1);
tail(a) + tail(head(a));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`recursive pair definitions are possible (tail): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair (1,a);
head(a) + head(tail(a));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove all ones on infinite list of ones and twos: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,pair(2,a));
const b = remove_all(1,a);
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove not found: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal (remove(2, list(1)),list(1));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove on infinite list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = pair(1,a);
const b = remove(1,a);
list_ref(b,200);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "remove(1, list(1));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove_all not found: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(remove_all(1, list(2, 3, 4)), list(2, 3, 4));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove_all: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(remove_all(1, list(1, 2, 3, 4, 1, 1, 1, 5, 1, 1, 6)), list(2, 3, 4, 5, 6));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`reverse: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(reverse(list(\\"string\\", \\"null\\", \\"undefined\\", \\"null\\", 123)), list(123, \\"null\\", \\"undefined\\", \\"null\\", \\"string\\"));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`tail of a 1 element list is null: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(list(1));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`tail works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "tail(pair(1, 'a string \\"\\"'));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "a string \\"\\"",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
